external/fv3fit/fv3fit/keras/_models/recurrent.py

```diff
@@ -0,0 +1,611 @@
+                Moisture can still go negative if there is a negative
+                physics tendency of moisture.
+            state_noise: amount of Gaussian noise to add to model prognostic state
+                before using it to predict tendencies
+        """
+        prognostic_variables = (
+            "air_temperature",
+            "specific_humidity",
+        )
+        self.given_tendency_names = (
+            "air_temperature_tendency_due_to_model",
+            "specific_humidity_tendency_due_to_model",
+        )
+        output_variables = ["dQ1", "dQ2"]
+        super(BPTTModel, self).__init__(
+            sample_dim_name, input_variables, output_variables,
+        )
+        self.input_packer = ArrayPacker(sample_dim_name, input_variables)
+        self.prognostic_packer = ArrayPacker(sample_dim_name, prognostic_variables)
+        self.train_timestep_seconds: Optional[float] = None
+        self.input_scaler = LayerStandardScaler()
+        self.prognostic_scaler = LayerStandardScaler()
+        self.tendency_scaler = LayerStandardScaler()
+        self.n_units = n_units
+        self.n_hidden_layers = n_hidden_layers
+        # self.tendency_ratio = tendency_ratio
+        self.kernel_regularizer = kernel_regularizer
+        self.train_batch_size = train_batch_size
+        self.optimizer = optimizer
+        self.activation = "relu"
+        self.train_keras_model: Optional[tf.keras.Model] = None
+        self.train_tendency_model: Optional[tf.keras.Model] = None
+        self.predict_keras_model: Optional[tf.keras.Model] = None
+        self.use_moisture_limiter = use_moisture_limiter
+        self.state_noise = state_noise
+        self._stepwise_output_metadata: List[DataArrayMetadata] = []
+
+    def build_for(self, X: xr.Dataset):
+        """
+        Given a dataset with [sample, time, z] and [sample, time] arrays
+        containing data in sample windows, fit the scalers and packers
+        and define the keras model (e.g. with the given window length).
+        Timestep is determined from the difference between the first two
+        times in the dataset.
+        """
+        if self.train_keras_model is not None:
+            raise RuntimeError("cannot build, model is already built!")
+        inputs = self.input_packer.to_array(X, is_3d=True)
+        state = self.prognostic_packer.to_array(X, is_3d=True)
+        self.input_scaler.fit(inputs)
+        self.prognostic_scaler.fit(state)
+        self.tendency_scaler.std = self.prognostic_scaler.std
+        self.tendency_scaler.mean = np.zeros_like(self.prognostic_scaler.mean)
+        time = X[BPTTModel.TIME_DIM_NAME]
+        self._train_timestep_seconds = (
+            time[1].values.item() - time[0].values.item()
+        ).total_seconds()
+        (
+            self.train_keras_model,
+            self.train_tendency_model,
+            self.predict_keras_model,
+        ) = self.build(len(time) - 1)
```

Comment:
I should do this at the start of fit if the train model is undefined, and rename this helper function to something like `fit_statistics` to reflect that it's setting the scaler and packer metadata.